File: daytona-terminal/backend/app/api/workspaces.py
"""Workspace API routes."""

from fastapi import APIRouter, HTTPException, status

from app.models.workspace import (
    Workspace,
    WorkspaceCreate,
    WorkspaceList,
    WorkspaceUpdate,
)
from app.services.daytona import DaytonaError, get_daytona_service

router = APIRouter(prefix="/workspaces", tags=["workspaces"])


@router.post("", response_model=Workspace, status_code=status.HTTP_201_CREATED)
async def create_workspace(request: WorkspaceCreate):
    """Create a new Daytona workspace."""
    try:
        service = get_daytona_service()
        workspace = await service.create_workspace(request)
        return workspace
    except DaytonaError as e:
        raise HTTPException(
            status_code=e.status_code or status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=e.message,
        )


@router.get("", response_model=WorkspaceList)
async def list_workspaces(page: int = 1, per_page: int = 20):
    """List all workspaces."""
    service = get_daytona_service()
    workspaces, total = await service.list_workspaces(page, per_page)
    return WorkspaceList(
        workspaces=workspaces,
        total=total,
        page=page,
        per_page=per_page,
    )


@router.get("/{workspace_id}", response_model=Workspace)
async def get_workspace(workspace_id: str):
    """Get a workspace by ID."""
    service = get_daytona_service()
    workspace = await service.get_workspace(workspace_id)
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workspace not found: {workspace_id}",
        )
    return workspace


@router.patch("/{workspace_id}", response_model=Workspace)
async def update_workspace(workspace_id: str, request: WorkspaceUpdate):
    """Update a workspace."""
    service = get_daytona_service()
    workspace = await service.update_workspace(workspace_id, request)
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workspace not found: {workspace_id}",
        )
    return workspace


@router.delete("/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workspace(workspace_id: str):
    """Delete a workspace."""
    service = get_daytona_service()
    try:
        success = await service.delete_workspace(workspace_id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Workspace not found: {workspace_id}",
            )
    except DaytonaError as e:
        raise HTTPException(
            status_code=e.status_code or status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=e.message,
        )


@router.post("/{workspace_id}/start", response_model=Workspace)
async def start_workspace(workspace_id: str):
    """Start a stopped workspace."""
    service = get_daytona_service()
    try:
        workspace = await service.start_workspace(workspace_id)
        if not workspace:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Workspace not found: {workspace_id}",
            )
        return workspace
    except DaytonaError as e:
        raise HTTPException(
            status_code=e.status_code or status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=e.message,
        )


@router.post("/{workspace_id}/stop", response_model=Workspace)
async def stop_workspace(workspace_id: str):
    """Stop a running workspace."""
    service = get_daytona_service()
    try:
        workspace = await service.stop_workspace(workspace_id)
        if not workspace:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Workspace not found: {workspace_id}",
            )
        return workspace
    except DaytonaError as e:
        raise HTTPException(
            status_code=e.status_code or status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=e.message,
        )


@router.post("/{workspace_id}/exec")
async def execute_command(
    workspace_id: str,
    command: str,
    working_dir: str | None = None,
    timeout: int = 60,
):
    """Execute a command in a workspace."""
    service = get_daytona_service()
    try:
        output, exit_code = await service.execute_command(
            workspace_id,
            command,
            working_dir=working_dir,
            timeout=timeout,
        )
        return {"output": output, "exit_code": exit_code}
    except DaytonaError as e:
        raise HTTPException(
            status_code=e.status_code or status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=e.message,
        )

File: daytona-terminal/backend/src/api/workspaces.ts
import type { FastifyPluginAsync } from 'fastify';
import { z } from 'zod';
import { daytonaService, DaytonaError } from '../services/daytona.service.js';
import { WorkspaceCreateSchema, WorkspaceUpdateSchema } from '../models/workspace.js';

export const workspacesRoutes: FastifyPluginAsync = async (fastify) => {
  // List workspaces
  fastify.get('/', async (request, reply) => {
    const { page = 1, perPage = 20 } = request.query as { page?: number; perPage?: number };
    const result = await daytonaService.listWorkspaces(page, perPage);
    return {
      workspaces: result.workspaces,
      total: result.total,
      page,
      perPage,
    };
  });

  // Create workspace
  fastify.post('/', async (request, reply) => {
    try {
      const body = WorkspaceCreateSchema.parse(request.body);
      const workspace = await daytonaService.createWorkspace(body);
      reply.code(201);
      return workspace;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Validation error', details: error.errors };
      }
      if (error instanceof DaytonaError) {
        reply.code(error.statusCode || 500);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Get workspace
  fastify.get('/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    const workspace = await daytonaService.getWorkspace(id);
    if (!workspace) {
      reply.code(404);
      return { error: `Workspace not found: ${id}` };
    }
    return workspace;
  });

  // Update workspace
  fastify.patch('/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    try {
      const body = WorkspaceUpdateSchema.parse(request.body);
      const workspace = await daytonaService.updateWorkspace(id, body);
      if (!workspace) {
        reply.code(404);
        return { error: `Workspace not found: ${id}` };
      }
      return workspace;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Validation error', details: error.errors };
      }
      throw error;
    }
  });

  // Delete workspace
  fastify.delete('/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    try {
      const success = await daytonaService.deleteWorkspace(id);
      if (!success) {
        reply.code(404);
        return { error: `Workspace not found: ${id}` };
      }
      reply.code(204);
      return;
    } catch (error) {
      if (error instanceof DaytonaError) {
        reply.code(error.statusCode || 500);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Start workspace
  fastify.post('/:id/start', async (request, reply) => {
    const { id } = request.params as { id: string };
    try {
      const workspace = await daytonaService.startWorkspace(id);
      if (!workspace) {
        reply.code(404);
        return { error: `Workspace not found: ${id}` };
      }
      return workspace;
    } catch (error) {
      if (error instanceof DaytonaError) {
        reply.code(error.statusCode || 500);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Stop workspace
  fastify.post('/:id/stop', async (request, reply) => {
    const { id } = request.params as { id: string };
    try {
      const workspace = await daytonaService.stopWorkspace(id);
      if (!workspace) {
        reply.code(404);
        return { error: `Workspace not found: ${id}` };
      }
      return workspace;
    } catch (error) {
      if (error instanceof DaytonaError) {
        reply.code(error.statusCode || 500);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Execute command
  fastify.post('/:id/exec', async (request, reply) => {
    const { id } = request.params as { id: string };
    const { command, workingDir, timeout } = request.body as {
      command: string;
      workingDir?: string;
      timeout?: number;
    };

    if (!command) {
      reply.code(400);
      return { error: 'Command is required' };
    }

    try {
      const result = await daytonaService.executeCommand(id, command, {
        workingDir,
        timeout,
      });
      return result;
    } catch (error) {
      if (error instanceof DaytonaError) {
        reply.code(error.statusCode || 500);
        return { error: error.message };
      }
      throw error;
    }
  });
};
